app/uk/gov/hmrc/gform/eval/EvaluationResults.scala

```diff
@@ -778,37 +778,42 @@ case class EvaluationResults(
   ) = {
     val concatValue = exprs
       .map { expr =>
-        val evalTypeInfo: TypeInfo = expr match {
-          case Add(_, _) | Multiply(_, _) | Subtraction(_, _) | Divide(_, _) =>
-            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
-          case DateCtx(_) => TypeInfo(expr, StaticTypeData(ExprType.dateString, None))
-          case IsNumberConstant(_) | PeriodExt(_, _) | UserCtx(UserField.Enrolment(_, _, Some(UserFieldFunc.Count))) |
-              Size(_, _) | CsvCountryCountCheck(_, _, _) =>
-            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
-          case DataRetrieveCtx(_, attribute) if attribute.`type` === ExprType.number =>
-            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
-          case DataRetrieveCount(_) =>
-            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
-          case Period(_, _) | PeriodValue(_) => TypeInfo(expr, StaticTypeData(ExprType.period, None))
-          case Typed(_, tpe)                 => TypeInfo(expr, StaticTypeData.from(tpe))
-          case DateFunction(_)               => TypeInfo(expr, StaticTypeData(ExprType.number, None))
-          case AuthCtx(AuthInfo.ItmpAddress) => TypeInfo(expr, StaticTypeData(ExprType.address, None))
-          case _                             => TypeInfo(expr, StaticTypeData(ExprType.string, None))
-        }
-
-        val stringResult = evalExpr(
-          evalTypeInfo,
-          recData,
-          booleanExprResolver,
-          evaluationContext
-        ).stringRepresentation(evalTypeInfo, evaluationContext.messages)
-
         expr match {
-          case Typed(_, ExplicitExprType.Sterling(_)) =>
-            TextFormatter.formatSterling(stringResult)
-          case Typed(_, ExplicitExprType.Number(fractionalDigits, roundingMode)) =>
-            TextFormatter.formatNumberWithPrecise(stringResult, fractionalDigits, roundingMode)
-          case _ => stringResult
+          case Constant(value) => value
```

Comment:
Not sure about matching the same expression multiple times in nested scopes.
I tried to refactor, treating  Constant(value) as other's expressions with ExprType.string. 
It compiles, not sure if it works though...
```
  private def evalConcat(
    exprs: List[Expr],
    recData: RecData[SourceOrigin.OutOfDate],
    booleanExprResolver: BooleanExprResolver,
    evaluationContext: EvaluationContext
  ) = {
    val concatValue = exprs
      .map((expr: Expr) =>
        expr match {
          case Add(_, _) | Multiply(_, _) | Subtraction(_, _) | Divide(_, _) =>
            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
          case DateCtx(_) => TypeInfo(expr, StaticTypeData(ExprType.dateString, None))
          case IsNumberConstant(_) | PeriodExt(_, _) | UserCtx(UserField.Enrolment(_, _, Some(UserFieldFunc.Count))) |
              Size(_, _) | CsvCountryCountCheck(_, _, _) =>
            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
          case DataRetrieveCtx(_, attribute) if attribute.`type` === ExprType.number =>
            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
          case DataRetrieveCount(_) =>
            TypeInfo(expr, StaticTypeData(ExprType.number, Some(Number())))
          case Period(_, _) | PeriodValue(_) => TypeInfo(expr, StaticTypeData(ExprType.period, None))
          case Typed(_, tpe)                 => TypeInfo(expr, StaticTypeData.from(tpe))
          case DateFunction(_)               => TypeInfo(expr, StaticTypeData(ExprType.number, None))
          case AuthCtx(AuthInfo.ItmpAddress) => TypeInfo(expr, StaticTypeData(ExprType.address, None))
          case _                             => TypeInfo(expr, StaticTypeData(ExprType.string, None))
        }
      )
      .map { (evalTypeInfo: TypeInfo) =>
        val stringResult = evalExpr(
          evalTypeInfo,
          recData,
          booleanExprResolver,
          evaluationContext
        ).stringRepresentation(evalTypeInfo, evaluationContext.messages)
        evalTypeInfo match {
          case ti @ TypeInfo(Typed(_, ExplicitExprType.Sterling(_)), _) =>
            TextFormatter.formatSterling(stringResult)
          case ti @ TypeInfo(Typed(_, ExplicitExprType.Number(fractionalDigits, roundingMode)), _) =>
            TextFormatter.formatNumberWithPrecise(stringResult, fractionalDigits, roundingMode)
          case _ => stringResult
        }
      }
      .mkString("")
    StringResult(concatValue)
  }
```

Reply:
The problem is `stringRepresentation` function removes all spaces. If you try the test [https://github.com/hmrc/gform-frontend/pull/1790/files#diff-e449ec8230d6c0f7379ac4c0a78e4b7ac09a2c00578f80ff4d57b630b6c8c964R515](https://github.com/hmrc/gform-frontend/pull/1790/files#diff-e449ec8230d6c0f7379ac4c0a78e4b7ac09a2c00578f80ff4d57b630b6c8c964R515), it fails.